File: flutter_ui/lib/models/tool.dart
class ToolEnvVar {
  final String key;
  final String description;
  final bool required;

  ToolEnvVar({required this.key, this.description = '', this.required = true});

  factory ToolEnvVar.fromJson(Map<String, dynamic> json) => ToolEnvVar(
        key: json['key'] ?? '',
        description: json['description'] ?? '',
        required: json['required'] ?? true,
      );

  Map<String, dynamic> toJson() => {
        'key': key,
        'description': description,
        'required': required,
      };
}

class Tool {
  final String id;
  final String name;
  final String description;
  final List<String> tags;
  final String installCommands;
  final String checkCommand;
  final List<ToolEnvVar> envVars;
  final String? authScript;
  final String? skillContent;
  final String version;
  final String author;
  final String? license;
  final String? sourceUrl;
  final bool enabled;
  final String? createdAt;
  final String? updatedAt;

  Tool({
    required this.id,
    required this.name,
    this.description = '',
    this.tags = const [],
    required this.installCommands,
    required this.checkCommand,
    this.envVars = const [],
    this.authScript,
    this.skillContent,
    this.version = '',
    this.author = '',
    this.license,
    this.sourceUrl,
    this.enabled = true,
    this.createdAt,
    this.updatedAt,
  });

  factory Tool.fromJson(Map<String, dynamic> json) => Tool(
        id: json['id'] ?? '',
        name: json['name'] ?? '',
        description: json['description'] ?? '',
        tags: List<String>.from(json['tags'] ?? []),
        installCommands: json['install_commands'] ?? '',
        checkCommand: json['check_command'] ?? '',
        envVars: (json['env_vars'] as List?)
                ?.map((e) => ToolEnvVar.fromJson(e))
                .toList() ??
            [],
        authScript: json['auth_script'],
        skillContent: json['skill_content'],
        version: json['version'] ?? '',
        author: json['author'] ?? '',
        license: json['license'],
        sourceUrl: json['source_url'],
        enabled: json['enabled'] ?? true,
        createdAt: json['created_at'],
        updatedAt: json['updated_at'],
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'description': description,
        'tags': tags,
        'install_commands': installCommands,
        'check_command': checkCommand,
        'env_vars': envVars.map((e) => e.toJson()).toList(),
        if (authScript != null) 'auth_script': authScript,
        if (skillContent != null) 'skill_content': skillContent,
        'version': version,
        'author': author,
        if (license != null) 'license': license,
        if (sourceUrl != null) 'source_url': sourceUrl,
        'enabled': enabled,
      };
}

File: crates/claw-models/src/tool.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// A provisioned CLI tool (e.g., az-cli, aws-cli) that can be installed into
/// Docker sandbox images and made available during job execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tool {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub tags: Vec<String>,
    /// Shell commands to install the tool (targets Docker/Debian).
    /// Multiline: each line is joined with ` && ` for Dockerfile RUN directives.
    pub install_commands: String,
    /// Command to check if tool is already installed (exit 0 = present).
    /// Example: "az --version" or "aws --version"
    pub check_command: String,
    /// Environment variables this tool needs at runtime (documentation + credential binding).
    #[serde(default)]
    pub env_vars: Vec<ToolEnvVar>,
    /// Optional login/auth commands run before claude -p in the container.
    /// These can reference credential env vars injected at runtime.
    #[serde(default)]
    pub auth_script: Option<String>,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub author: String,
    #[serde(default)]
    pub license: Option<String>,
    #[serde(default)]
    pub source_url: Option<String>,
    #[serde(default = "default_true")]
    pub enabled: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Declares an environment variable a tool needs at runtime.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolEnvVar {
    /// Env var name (e.g., "AZURE_CLIENT_ID").
    pub key: String,
    /// Human-readable description.
    #[serde(default)]
    pub description: String,
    /// Whether this variable is required for the tool to function.
    #[serde(default = "default_true")]
    pub required: bool,
}

fn default_true() -> bool {
    true
}

/// Request body for creating a new tool via the API.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateToolRequest {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub tags: Vec<String>,
    pub install_commands: String,
    pub check_command: String,
    #[serde(default)]
    pub env_vars: Vec<ToolEnvVar>,
    #[serde(default)]
    pub auth_script: Option<String>,
}
